src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/ReclusterService.kt

```diff
@@ -5,39 +5,81 @@ import org.springframework.beans.factory.annotation.Value
 import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType.CPR_UUID_RECLUSTER_NEEDS_ATTENTION
 
 @Component
 class ReclusterService(
+  private val matchService: MatchService,
   private val telemetryService: TelemetryService,
   @Value("\${retry.delay}") private val retryDelay: Long,
 ) {
 
-  fun recluster(personEntity: PersonEntity) = runBlocking {
+  fun recluster(personKeyEntity: PersonKeyEntity) = runBlocking {
     runWithRetry(MAX_ATTEMPTS, retryDelay, ENTITY_RETRY_EXCEPTIONS) {
-      handleRecluster(personEntity)
+      handleRecluster(personKeyEntity)
     }
   }
 
-  private fun handleRecluster(personEntity: PersonEntity) {
+  private fun handleRecluster(personKeyEntity: PersonKeyEntity) {
     when {
-      clusterNeedsAttention(personEntity.personKey) -> telemetryService.trackPersonEvent(
+      clusterNeedsAttention(personKeyEntity) -> telemetryService.trackEvent(
         CPR_UUID_RECLUSTER_NEEDS_ATTENTION,
-        personEntity,
         mapOf(
-          EventKeys.UUID to personEntity.personKey?.personId.toString(),
+          EventKeys.UUID to personKeyEntity.personId.toString(),
         ),
       )
-      clusterHasOneRecord(personEntity.personKey) -> {} // CPR-437
-      else -> {} // CPR-439
+      clusterHasOneRecord(personKeyEntity) -> {} // CPR-437
+      else -> handleMultipleRecordsInCluster(personKeyEntity)
     }
   }
 
+  private fun handleMultipleRecordsInCluster(personKeyEntity: PersonKeyEntity) {
+    checkClusterRecordsMatch(personKeyEntity)
+    // CPR-452
+  }
+
+  private fun checkClusterRecordsMatch(personKeyEntity: PersonKeyEntity): List<PersonEntity> {
+    val recordsInClusterMatched: MutableList<PersonEntity> = mutableListOf()
+    val recordsInClusterNotMatched: MutableList<PersonEntity> = personKeyEntity.personEntities.toMutableList()
+
+    personKeyEntity.personEntities.forEach { personEntity ->
+      when {
+        recordsInClusterMatched.contains(personEntity) -> return@forEach
+      }
+      val (matchedRecords, notMatchedRecords) = matchRecordAgainstCluster(personEntity, personKeyEntity.personEntities)
+      when {
+        matchedRecords.isNotEmpty() -> {
+          addAllIfNotPresent(recordsInClusterMatched, matchedRecords + personEntity)
+          removeAllIfPresent(recordsInClusterNotMatched, notMatchedRecords + personEntity)
+        }
+      }
+    }
+
+    return recordsInClusterNotMatched.toList()
+  }
+
+  private fun matchRecordAgainstCluster(recordToMatch: PersonEntity, personEntities: List<PersonEntity>): Pair<List<PersonEntity>, List<PersonEntity>> {
+    val recordsToMatch = personEntities.filterNot { it == recordToMatch }
+    val matchedRecords = matchService.findHighConfidenceMatches(recordsToMatch, PersonSearchCriteria.from(recordToMatch)).map { it.candidateRecord }
+    val unmatchedRecords = recordsToMatch.filter { matchedRecord -> matchedRecords.map { it.id }.contains(matchedRecord.id).not() }
+    return Pair(matchedRecords, unmatchedRecords)
+  }
```

Comment:
```suggestion
   private fun matchRecordAgainstCluster(recordToMatch: PersonEntity, personEntities: List<PersonEntity>): Pair<List<PersonEntity>, List<PersonEntity>> {
    val recordsToMatch = personEntities.filterNot { it == recordToMatch }
    return matchService.findHighConfidenceMatches(recordsToMatch, PersonSearchCriteria.from(recordToMatch))
      .map { it.candidateRecord }
      .partition { !recordsToMatch.contains(it) }
  }
```

Is this any better? The tests pass....

Reply:
Updated 👍 , need to be partitioned on the record match as the matchservice only returns high confidence matches.